zipline/finance/commission.py

```diff
@@ -79,56 +204,114 @@ def __init__(self,
-            return max(self.min_trade_cost, additional_commission)
+    def __init__(self,
+                 cost_per_contract,
+                 exchange_fee,
+                 min_trade_cost=DEFAULT_MINIMUM_COST_PER_TRADE):
+        if isinstance(cost_per_contract, int):
+            cost_per_contract = float(cost_per_contract)
+        if isinstance(exchange_fee, int):
+            exchange_fee = float(exchange_fee)
+        self.cost_per_contract = cost_per_contract
+        self.exchange_fee = exchange_fee
+        self.min_trade_cost = min_trade_cost
+
+    def __repr__(self):
+        if isinstance(self.cost_per_contract, float):
+            cost_per_contract = self.cost_per_contract
         else:
-            # we've already paid some commission, so figure out how much we
-            # would be paying if we only counted per share.
-            per_share_total = \
-                (order.filled * self.cost_per_share) + additional_commission
+            cost_per_contract = '[varies]'
 
-            if per_share_total < self.min_trade_cost:
-                # if we haven't hit the minimum threshold yet, don't pay
-                # additional commission
-                return 0
-            else:
-                # we've exceeded the threshold, so pay more commission.
-                return per_share_total - order.commission
+        if isinstance(self.exchange_fee, float):
+            exchange_fee = self.exchange_fee
+        else:
+            exchange_fee = '[varies]'
+
+        return (
+            '{class_name}(cost_per_contract={cost_per_contract}, '
+            'exchange_fee={exchange_fee}, min_trade_cost={min_trade_cost})'
+            .format(
+                class_name=self.__class__.__name__,
+                cost_per_contract=cost_per_contract,
+                exchange_fee=exchange_fee,
+                min_trade_cost=self.min_trade_cost,
+            )
+        )
+
+    def calculate(self, order, transaction):
+        if isinstance(self.cost_per_contract, float):
+            cost_per_contract = self.cost_per_contract
+        else:
+            # Cost per contract is a dictionary. If the user's dictionary does
+            # not provide a commission cost for a certain contract, fall back
+            # on the default.
+            root_symbol = order.asset.root_symbol
+            cost_per_contract = self.cost_per_contract.get(
+                root_symbol, DEFAULT_PER_CONTRACT_COST,
+            )
+
+        if isinstance(self.exchange_fee, float):
+            exchange_fee = self.exchange_fee
+        else:
+            # Exchange fee is a dictionary. If the user's dictionary does not
+            # provide an exchange fee for a certain contract, fall back on the
+            # default.
+            root_symbol = order.asset.root_symbol
+            exchange_fee = self.exchange_fee.get(
+                root_symbol, FUTURE_EXCHANGE_FEES_BY_SYMBOL[root_symbol],
+            )
```

Comment:
The logic here that differs according to the type of `exchange_fee` and `cost_per_contract` seems well suited for something like multiple dispatch, but I'm not sure how that would work if these values aren't parameters to the method.